onnx/helper.py

```diff
@@ -617,6 +618,35 @@ def float32_to_float8e5m2(  # noqa: PLR0911
         raise NotImplementedError("fn and uz must be both False or True.")
 
 
+"""Convert an array of float32 value to a 4bit data-type and pack every two concecutive elements in a byte.
+
+See :ref:`onnx-detail-int4` for technical details.
+
+Args:
+    farray: array of float to convert and pack
+    signed: Whether the 4 bit variant is signed or unsigned
+
+Returns:
+    Packed array with size `ceil(farray.size/2)` (single dimension).
+"""
+
+
+def pack_float32_to_4bit(farray: np.ndarray, signed: bool) -> np.ndarray:
```

Comment:
is farray float array? I would avoid abbreviations and always spell out the words (with exceptions like "op").